slides/advanced-01-tuning-hyperparameters.qmd

```diff
@@ -0,0 +1,696 @@
+* The rate that each tree adapts from previous iterations (`learn_rate` in $(0, \infty]$, usual maximum is 0.1).
+* The number of iterations of boosting where _no improvement_ was shown before stopping (`stop_iter` in $[1, $`trees`$]$)
+
+## Boosted Tree Tuning Parameters
+
+TBH it is usually not difficult to optimize these models. 
+
+<br>
+
+Often, there are multiple _candidate_ tuning parameter combinations that have very good results. 
+
+<br>
+
+To demonstrate simple concepts, we'll look at optimizing the number of trees in the ensemble (between 1 and 100) and the learning rate ($10^{-5}$ to $10^{-1}$).
+
+## Boosted Tree Tuning Parameters `r hexes(c("workflows", "parsnip", "bonsai"))`
+
+We'll need to load the bonsai package. This has the information needed to use lightgbm
+
+```{r}
+#| label: boot-spec
+
+library(bonsai)
+lgbm_spec <- 
+  boost_tree(trees = tune(), learn_rate = tune()) %>% 
+  set_mode("regression") %>% 
+  set_engine("lightgbm")
+
+lgbm_wflow <- workflow(hotel_rec, lgbm_spec)
+```
+
+
+## Optimize tuning parameters
+
+The main two strategies for optimization are:
+
+. . .
+
+- **Grid search** 💠 which tests a pre-defined set of candidate values
+
+- **Iterative search** 🌀 which suggests/estimates new values of candidate parameters to evaluate
+
+
+## Grid search
+
+A small grid of points trying to minimize the error via learning rate: 
+
+```{r}
+#| label: small-grid-demo
+#| echo: false
+#| fig-align: center
+#| out-width: 60%
+
+knitr::include_graphics("images/small_init.svg")
+```
+
+
+## Grid search
+
+In reality we would probably sample the space more densely: 
+
+```{r}
+#| label: grid-demo
+#| echo: false
+#| fig-align: center
+#| out-width: 60%
+
+knitr::include_graphics("images/grid_points.svg")
+```
+
+
+## Sequential Search
+
+We could start with a few points and search the space:
+
+```{r}
+#| label: seq-demo
+#| echo: false
+#| fig-align: center
+#| out-width: 60%
+
+knitr::include_graphics("images/anime_seq.gif")
+```
+
+# Grid Search
+
+## Parameters
+
+-   The tidymodels framework provides pre-defined information on tuning parameters (such as their type, range, transformations, etc).
+
+-   The `extract_parameter_set_dials()` function extracts these tuning parameters and the info.
+
+::: fragment
+#### Grids
+
+-   Create your grid manually or automatically.
+
+-   The `grid_*()` functions can make a grid.
+:::
+
+::: notes
+Most basic (but very effective) way to tune models
+:::
+
+## Create a grid `r hexes(c("dials", "workflows"))`
+
+```{r get-param}
+#| tidy: false
+lgbm_wflow %>% 
+  extract_parameter_set_dials()
+
+# Individual functions: 
+trees()
+learn_rate()
+```
+
+::: fragment
+A parameter set can be updated (e.g. to change the ranges).
+:::
+
+## Create a grid `r hexes(c("dials", "workflows"))`
+
+::: columns
+::: {.column width="50%"}
+```{r}
+#| label: get-grid 
+set.seed(12)
+grid <- 
+  lgbm_wflow %>% 
+  extract_parameter_set_dials() %>% 
+  grid_latin_hypercube(size = 25)
+
+grid
+```
```

Comment:
I think it would be nice to show a visual of what a *space-filling design* looks like.  Like the one shown on the **The results** slide